hoomd/hpmc/ExternalFieldHarmonic.h

```diff
@@ -0,0 +1,482 @@
+            dE += new_E - old_E;
+            }
+
+#ifdef ENABLE_MPI
+        if (this->m_sysdef->isDomainDecomposed())
+            {
+            MPI_Allreduce(MPI_IN_PLACE,
+                          &dE,
+                          1,
+                          MPI_HOOMD_SCALAR,
+                          MPI_SUM,
+                          m_exec_conf->getMPICommunicator());
+            }
+#endif
+
+        return dE;
+        }
+
+    /** This function gets exported to python to act as a property
+     *
+     * The include_translational and include_rotational arguments allow the purely rotational
+     * or purely orientational energies to be calculated separately
+     */
+    Scalar getEnergy(uint64_t timestep, bool include_translational, bool include_rotational)
+        {
+        Scalar energy = Scalar(0.0);
+        // access particle data and system box
+        ArrayHandle<Scalar4> h_postype(m_pdata->getPositions(),
+                                       access_location::host,
+                                       access_mode::read);
+        ArrayHandle<Scalar4> h_orient(m_pdata->getOrientationArray(),
+                                      access_location::host,
+                                      access_mode::read);
+        for (unsigned int i = 0; i < m_pdata->getN(); i++)
+            {
+            vec3<Scalar> position(h_postype.data[i]);
+            quat<Scalar> orientation(h_orient.data[i]);
+            energy += calcE(i, position, orientation, include_translational, include_rotational);
+            }
+
+#ifdef ENABLE_MPI
+        if (this->m_sysdef->isDomainDecomposed())
+            {
+            MPI_Allreduce(MPI_IN_PLACE,
+                          &energy,
+                          1,
+                          MPI_HOOMD_SCALAR,
+                          MPI_SUM,
+                          m_exec_conf->getMPICommunicator());
+            }
+#endif
+
+        return energy;
+        } // end getEnergy(uin64_t)
+
+    //! Calculate the change in energy from moving a single particle with tag = index
+    double energydiff(const unsigned int& index,
+                      const vec3<Scalar>& position_old,
+                      const Shape& shape_old,
+                      const vec3<Scalar>& position_new,
+                      const Shape& shape_new)
+        {
+        double old_U = calcE(index, position_old, shape_old, true, true),
+               new_U = calcE(index, position_new, shape_new, true, true);
+        return new_U - old_U;
+        }
+
+    protected:
+    //! Calculate the energy associated with the deviation of a single particle from its reference
+    //! position
+    Scalar calcE_trans(const unsigned int& index, const vec3<Scalar>& position)
+        {
+        ArrayHandle<unsigned int> h_tags(m_pdata->getTags(),
+                                         access_location::host,
+                                         access_mode::read);
+        int3 dummy = make_int3(0, 0, 0);
```

Comment:
This is only used in box wap. Go ahead and create it there.

Reply:
Do you just mean move it to the line before the `box.wrap()` line?